### DISCUSSION TITLE
Sharing field implementation with different field names

### DISCUSSION THREAD

**Discussion by danielbuechele** [hayes/pothos#813] [Q&A]
A few of my nodes expose access to an image asset type, which is defined as follows:
```
PixelImage = {
  uri: String!
  copyright: String
  width: Int!
  height Int!
}
```

However,  the name of this field is different for my nodes. On some it might be called `image`, on others `preview` or `thumbnail`. What is a good way of sharing the implementation of this field and args, while allowing different field names? I could just have a function which returns the arg and another one which resolves the field, but now args and resolver are disconnected, and making this type safe is hard.
```
preview: t.field({
    type: PixelImage,
    args: pixelImageArgs(t),
    resolve: (root) => resolvePixelImage(root.preview),
}),
```

I ideally I'd have something like, which generates the entire field, but again I'm running into trouble making this type-safe.
```
preview: pixelImageField(t, 'preview'),
```

Any other alternatives how to do this? If all of the fields had the same name, it would be easy to implement using an interface, but because they have different names, this isn't really possible?

**Comment by hayes** [comment]
I think you just need to make the parent shape a generic in your helper, and type the field name arg as keyof ParentShape

**Reply by danielbuechele** [reply]
Yeah, makes sense. I wasn't sure if the helper function is the way to go, but it seems to be the best approach. Still struggling to type the helper correctly, but that's unrelated to pothos and is more a problem between TypeScript and me!

This is what I currently have, but obviously I'd love to get away from the `any` and `string` types and would want this to be somehow dynamically "inherited" from the parent as a generic.
```
function resolvePixelImage(t: PrismaObjectFieldBuilder<any, any>, field: string) {}
```

Anyway, thanks for the help already.

**Reply by hayes** [reply]
```ts
const Image = builder.objectRef<{ url: string }>('Image').implement({});

builder
  .objectRef<{ id: number; title: string; image: { url: string }; preview: { url: string } }>(
    'Post',
  )
  .implement({
    fields: (t) => ({
      id: t.exposeID('id'),
      title: t.exposeString('title'),
      image: loadImage(t, 'image'),
      preview: loadImage(t, 'preview'),
    }),
  });

function loadImage<Types extends SchemaTypes, F extends string>(
  t: PothosSchemaTypes.FieldBuilder<Types, { [K in F]: { url: string } }, 'Object'>,
  field: F,
) {
  return t.field({
    type: Image,
    nullable: false,
    resolve: (parent) => parent[field],
  });
}
```
